File: ios/PinecilTime/IronOSUUIDs.swift
//
//  IronOSUUIDs.swift
//  PinecilTime
//

import CoreBluetooth
import Foundation

enum IronOSUUIDs {

    // MARK: - Bulk Data Service (for discovery)
    static let bulkDataService = CBUUID(string: "9EAE1000-9D0D-48C5-AA55-33E27F9BC533")
    static let bulkLiveData = CBUUID(string: "9EAE1001-9D0D-48C5-AA55-33E27F9BC533")
    static let buildID = CBUUID(string: "9EAE1003-9D0D-48C5-AA55-33E27F9BC533") // Firmware build version
    static let deviceSerial = CBUUID(string: "9EAE1004-9D0D-48C5-AA55-33E27F9BC533") // MAC address

    // MARK: - Live Data Service
    static let liveDataService = CBUUID(string: "D85EF000-168E-4A71-AA55-33E27F9BC533")
    static let liveTemp = CBUUID(string: "D85EF001-168E-4A71-AA55-33E27F9BC533")
    static let setpointRead = CBUUID(string: "D85EF002-168E-4A71-AA55-33E27F9BC533")
    static let dcInput = CBUUID(string: "D85EF003-168E-4A71-AA55-33E27F9BC533")
    static let handleTemp = CBUUID(string: "D85EF004-168E-4A71-AA55-33E27F9BC533")
    static let powerLevel = CBUUID(string: "D85EF005-168E-4A71-AA55-33E27F9BC533")
    static let powerSource = CBUUID(string: "D85EF006-168E-4A71-AA55-33E27F9BC533")
    static let tipResistance = CBUUID(string: "D85EF007-168E-4A71-AA55-33E27F9BC533")
    static let uptime = CBUUID(string: "D85EF008-168E-4A71-AA55-33E27F9BC533")
    static let lastMovement = CBUUID(string: "D85EF009-168E-4A71-AA55-33E27F9BC533")
    static let maxTemp = CBUUID(string: "D85EF00A-168E-4A71-AA55-33E27F9BC533")
    static let rawTip = CBUUID(string: "D85EF00B-168E-4A71-AA55-33E27F9BC533")
    static let hallSensor = CBUUID(string: "D85EF00C-168E-4A71-AA55-33E27F9BC533")
    static let operatingMode = CBUUID(string: "D85EF00D-168E-4A71-AA55-33E27F9BC533")
    static let estimatedWatts = CBUUID(string: "D85EF00E-168E-4A71-AA55-33E27F9BC533")

    // MARK: - Settings Service
    static let settingsService = CBUUID(string: "F6D80000-5A10-4EBA-AA55-33E27F9BC533")
    static let saveSettings = CBUUID(string: "F6D7FFFF-5A10-4EBA-AA55-33E27F9BC533")
    static let resetSettings = CBUUID(string: "F6D7FFFE-5A10-4EBA-AA55-33E27F9BC533")
    
    // Helper to generate setting UUID from index
    static func settingUUID(index: UInt16) -> CBUUID {
        let hexString = String(format: "F6D7%04X-5A10-4EBA-AA55-33E27F9BC533", index)
        return CBUUID(string: hexString)
    }
    
    // Helper to extract setting index from UUID
    static func settingIndex(from uuid: CBUUID) -> UInt16? {
        let uuidString = uuid.uuidString.uppercased()
        guard uuidString.hasPrefix("F6D7") && uuidString.hasSuffix("-5A10-4EBA-AA55-33E27F9BC533") else {
            return nil
        }
        let hexIndex = String(uuidString.prefix(8).suffix(4))
        return UInt16(hexIndex, radix: 16)
    }
}

File: android/app/src/main/java/com/tinkcil/data/ble/IronOSUUIDs.kt
package com.tinkcil.data.ble

import java.util.UUID

object IronOSUUIDs {
    // Services
    val BULK_DATA_SERVICE: UUID = UUID.fromString("9eae1000-9d0d-48c5-aa55-33e27f9bc533")
    val LIVE_DATA_SERVICE: UUID = UUID.fromString("d85ef000-168e-4a71-aa55-33e27f9bc533")
    val SETTINGS_SERVICE: UUID = UUID.fromString("f6d80000-5a10-4eba-aa55-33e27f9bc533")

    // Bulk data characteristics
    val BULK_LIVE_DATA: UUID = UUID.fromString("9eae1001-9d0d-48c5-aa55-33e27f9bc533")
    val BUILD_ID: UUID = UUID.fromString("9eae1003-9d0d-48c5-aa55-33e27f9bc533")
    val DEVICE_SERIAL: UUID = UUID.fromString("9eae1004-9d0d-48c5-aa55-33e27f9bc533")

    // Settings control
    val SAVE_SETTINGS: UUID = UUID.fromString("f6d7ffff-5a10-4eba-aa55-33e27f9bc533")
    val RESET_SETTINGS: UUID = UUID.fromString("f6d7fffe-5a10-4eba-aa55-33e27f9bc533")

    // Setting characteristic by index
    fun settingCharacteristic(index: Int): UUID =
        UUID.fromString("f6d7%04x-5a10-4eba-aa55-33e27f9bc533".format(index))

    // All setting indices used by the app
    val SETTING_INDICES = intArrayOf(0, 1, 2, 6, 7, 11, 13, 14, 17, 22, 24, 25, 26, 27, 28, 33, 34)
}
